my-sms/spec/controllers/students_controller_spec.rb

```diff
@@ -23,119 +25,115 @@
+    subject { get :show, { id: student.to_param } }
+    it 'returns a success response' do
+      should be_success
+      should render_template('show')
     end
   end
 
-  describe "GET #new" do
-    it "returns a success response" do
-      get :new, {}, valid_session
-      expect(response).to be_success
+  describe 'GET #new' do
+    subject { get :new }
+    it 'returns a success response' do
+      should be_success
+      should render_template('new')
     end
   end
 
-  describe "GET #edit" do
-    it "returns a success response" do
-      student = Student.create! valid_attributes
-      get :edit, {:id => student.to_param}, valid_session
-      expect(response).to be_success
+  describe 'GET #edit' do
+    subject { get :edit, { id: student.to_param } }
+    it 'returns a success response' do
+      should be_success
+      should render_template('edit')
     end
   end
 
-  describe "POST #create" do
-    context "with valid params" do
-      it "creates a new Student" do
-        expect {
-          post :create, {:student => valid_attributes}, valid_session
-        }.to change(Student, :count).by(1)
+  describe 'POST #create' do
+    subject { post :create, { student: student } }
+
+    context 'with valid params' do
+      let(:student) { FactoryBot.attributes_for(:student) }
+
+      it 'creates a new Student' do
+        expect { subject }.to change(Student, :count).by(1)
       end
 
-      it "redirects to the created student" do
-        post :create, {:student => valid_attributes}, valid_session
-        expect(response).to redirect_to(Student.last)
+      it 'redirects to the created student' do
+        should be_redirect
+        should redirect_to(Student.last)
       end
     end
 
-    context "with invalid params" do
+    context 'with invalid params' do
+      let(:student) { FactoryBot.attributes_for(:student, first_name: nil) }
+
       it "returns a success response (i.e. to display the 'new' template)" do
-        post :create, {:student => invalid_attributes}, valid_session
-        expect(response).to be_success
+        should be_success
+        should render_template('new')
       end
     end
   end
 
-  describe "PUT #update" do
-    context "with valid params" do
-      let(:new_attributes) {
-        skip("Add a hash of attributes valid for your model")
-      }
+  describe 'PUT #update' do
+    let(:student) { FactoryBot.create(:student) }
+    subject { put :update, { id: student.to_param, student: new_attributes } }
+
+    context 'with valid params' do
+      let(:new_attributes) { FactoryBot.attributes_for(:student) }
 
-      it "updates the requested student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => new_attributes}, valid_session
+      it 'updates the requested student' do
+        subject
         student.reload
-        skip("Add assertions for updated state")
+        attrs = %i[first_name last_name gender]
```

Comment:
So I think this test works, but I think we could clean it up a little. We generally try to do assignment outside of the test,  which helps keep the actual test blocks smaller and more contained to act, and assert, with any arranging happening outside that block.

Another potential issue is that as you're not using something like Faker to randomise your student's attributes, you're not actually guaranteeing that your update works, as the attributes you're passing in, so I would recommend setting the values either in the initial record, or the update params to ones that are explicitly not the default for this factory.

What do you think of the following:

```ruby
# Set the initial values to whatever you want, as long as they don't match the factory defaults

let!(:student) { FactoryBot.create(:student, first_name: 'foo', last_name: 'bar', gender: 'baz') }
let(:new_attributes) { FactoryBot.attributes_for(:student).stringify_keys }
let(:attr_names) { %w[first_name last_name gender] }
# Reload may or may not be necessary, recommend experimenting with that
let(:student_attrs) { student.reload.attributes.fetch_values(*attr_names) }
let(:new_attr_values) { new_attributes.fetch_values(*attr_names) }

it 'updates the requested student' do
  subject
  student.reload
  expect(student_attrs).to eq new_attr_values
end
```

For completeness, you could also consider the alternative assertion:

```ruby
it 'updates the requested student' do
  expect { subject }.to change { student_attrs }.to(new_attr_values)
end
```

Note that `expect` and `change` are passed a block in this case, so they can be evaluated at the correct times.

Reply:
I updated with your suggestions in 7b355989292cab5b4ebc1d4043e89b841347e435
I also experimented with reload and it turns out you don't need it. I replaced the block with the just the change expectation, however, I couldn't leave student_attrs in the let method as it will not re-evaluate the cache before and after the change. Therefore I've moved the attributes to check directly into the change block. I'm not sure if that's the best option here but it was cleaner and more idiomatic AFAIK than my other solution (putting a proc inside let and calling .call in the change block).

Thanks for stringify_keys, didn't know that existed.
Also the use of let! vs let was also something I hadn't considered.

Reply:
Actually, this broke later on and I had to add reload back in. I guess if you aren't careful the ActiveModel objects aren't deterministic?

Reply:
Yeah, the distinction between `let` and `let!` is a really important one, and useful to avoid needing to reference a variable just to create a record (which I might have suggested in another comment, oops).

Without the `reload`, the instance of the record is loaded into memory and rails doesn't know to reload from the database of this record is updated from your controller action, so the manual reload is necessary.

Because the controller is fetching the record separately, it's not making changes to the instance you've got in your test. You can see this more clearly by checking `student.object_id` in your `it` block vs the one fetched in the controller.

As an aside this is a really common gotcha when it comes to stubbing/mocking, where you'll add something like `allow(student).to receive(:first_name).and_return('foo')`, but because the instance in the controller is fetched separately, the stub isn't applied to the controller's `student`, so it returns the actual name on the record. In this case you need to either also stub the model's `find` to return your specific instance, or use `any_instance_of`.

I hope I've actually answered your question here. Let me know if I haven't.

Reply:
Thanks that's good to know.